File: ii-service/src/main/java/com/manymonkeys/service/cinema/MovieService.java
package com.manymonkeys.service.cinema;

import com.manymonkeys.model.cinema.*;
import com.manymonkeys.service.exception.NotFoundException;

import java.util.Map;

/**
 * @author Ilya Pimenov
 * @author Anton Chebotaev
 *         Owls Proprietary
 */
public interface MovieService {

    Movie createMovie(Movie movie);

    Movie load(String name, Long year) throws NotFoundException;

    Movie load(String service, String externalId) throws NotFoundException;

    Map<Movie, Double> getMostLike(Movie movie) throws NotFoundException;

    Movie setDescription(Movie movie, String description) throws NotFoundException;

    Movie setTranslateName(Movie movie, String translateName, Boolean index);

    Movie setExternalId(Movie movie, String service, String externalId) throws NotFoundException;

    Movie setAkaName(Movie movie, String akaName, Boolean index);

    Movie addPerson(Movie movie, Person person, Role role) throws NotFoundException;

    Movie addKeyword(Movie movie, Keyword keyword) throws NotFoundException;

    Boolean hasKeyword(Movie movie, Keyword keyword) throws NotFoundException;

    Movie addTagline(Movie movie, String tagline);

    Movie addGenre(Movie movie, Genre genre) throws NotFoundException;

}

File: ii-service/src/main/scala/com/manymonkeys/research/service/cinema/MovieService.scala
package com.manymonkeys.research.service.cinema

import com.manymonkeys.model.cinema._

/**
* @author Ilya Pimenov
*         Owls Proprietary
*/
trait MovieService {

  def createMovie(movie: Movie): Movie

  def loadByName(name: String, year: Long): Movie

  def getMostLike(movie: Movie): Map[Movie, Double]

  def createOrUpdateDescription(movie: Movie, description: String): Movie

  def loadByExternalId(service: String, externalId: String): Movie

  def addPerson(movie: Movie, person: Person, role: Role): Movie

  def addKeyword(movie: Movie, keyword: Keyword): Movie

  def hasKeyword(movie: Movie, keyword: Keyword): Boolean

  def addTagline(movie: Movie, tagline: String): Movie

  def setAkaName(movie: Movie, akaName: String): Movie

  def setTranslateName(movie: Movie, translateName: String): Movie

  def addExternalId(movie: Movie, service: String, externalId: String): Movie

}

